### ISSUE TITLE
Shadowmap shader only works in 640x480

### ISSUE THREAD

**Issue by markcwm** [markcwm/openb3dmax.help#3]
Hi,

for anyone interested in shader-based textured shadows, Angros just shared FB example usage of his CreateShadowMap function in Openb3d Plus, so I ported the shadowmap code to Blitzmax, see [shadowmap.bmx](https://github.com/markcwm/openb3d.docs/blob/master/shaders/shadowmap.bmx).

But sadly the main issue is that it only works in 640x480 resolution, if I try 320x240 or 800x600 I get the same result, the shadow seems to be rendered over the whole receiver mesh. 

And the same thing happens when I change the shadowmap texture size from 512x512 to 256 or 1024. 

Here is the [frag shader](https://github.com/markcwm/openb3d.docs/blob/master/glsl/shadowmap.frag.glsl) and here is the [vert shader](https://github.com/markcwm/openb3d.docs/blob/master/glsl/shadowmap.vert.glsl).

**Comment by blitzcoder** [comment]
Hey Mark,

Just stumbled upon this and trying to test it with a clean installation of vanilla bmx 1.50

```
Compile Error: Unable to convert from 'Float Array Ptr' to 'Float Ptr'
```

**Comment by markcwm** [comment]
Hi Ron,
yes I don't really test with Bmx 1.50 much now, but I think it just didn't need the VarPtr, not sure if VarPtr is needed for NG. Note that this requires GLSL 1.40 (GL 3.1).

**Comment by blitzcoder** [comment]
I see. getting a black cube on a blue background though when varptr is removed..

can you share the link or copy of the FB code?

![Untitled-1](https://user-images.githubusercontent.com/57479219/185281899-7708c5c2-5aa0-4f70-854f-4d1843ecf8fb.png)

**Comment by markcwm** [comment]
Hi Ron,

actually, I didn't test that fix properly so it wasn't right to remove VarPtr, glUniformMatrix4fv needs a float ptr and VarPtr takes that var and finds it's pointer, but you're also supposed to dereference a Blitzmax array with [0], NG let me away with that but not 1.50. The code below shows that biasMatrix is a 4x4 float array sent as a dereferenced array pointer, also CameraProjMatrix is a 4x4 float array, this is dereferenced in C++ by "return &proj_mat[0];" and then sent simply as a float ptr, it's not really tested anywhere else but I think that's OK.

`Local biasMatrix:Float[]=[0.5, 0.0, 0.0, 0.0,0.0, 0.5, 0.0, 0.0,0.0, 0.0, 0.5, 0.0,0.5, 0.5, 0.5, 1.0]
...
glUniformMatrix4fv(glGetUniformLocation(prog,"biasMatrix"), 1, 0, Varptr biasMatrix[0])
glUniformMatrix4fv(glGetUniformLocation(prog,"projMatrix"), 1, 0, CameraProjMatrix(CamLight))`

OT: I can't actually test this as I lost my laptop and haven't had time to set up my new one, Windows 10 has been very slow on my new laptop and I'm not sure what to do, I've tried disabling Cortana and all the bloatware including Aero but it's still very slow, have you any suggestions there? Thanks.

**Comment by markcwm** [comment]
There is no FB code, the original code was by Kronos in Blitzmax but with his own wrapper. I then ported the CreateShadowMap function.

> Got it working once I figured out every shadow receiver needs the texture applied to it. I am getting multiple shadows duplicated across the plane surface though.
I think you possibly need to change

> ShadowSampler=CreateTexture(512,512,1,1);
to
ShadowSampler=CreateTexture(512,512,1+16+32,1);
to clamp the texture in the function CreateShadowMap

> Would also be nice if we could define the size of the shadowmap through the function

```
Include "openb3d.bmx"

'wrapper stuff
Global libPlus:Int = LoadLibraryA("ob3dPlus_32.dll")
Global CreateShadowMap:Int(light:Int) "C" = getprocaddress(libPlus, "CreateShadowMap")
Global UpdateShadows() "C" = getprocaddress(libPlus, "UpdateShadows")


Graphics3D(1024,768)

camera = CreateCamera()
CameraClsColor(camera, 0, 0, 255)

cube = CreateCube()
EntityColor(cube, 255, 0, 0)
MoveEntity(cube,0,5,0)

anothercube=CreateCube()
MoveEntity(cube,1,0,0)



plane = CreateCube()
MoveEntity(plane,0,-4,0)
ScaleEntity(plane, 100, 1, 100)


light = CreateLight()

PositionEntity(light, 0, 20, 0)
PointEntity(light, cube)
MoveEntity(camera, -2, 0, -5)

shadow=CreateShadowmap(light)

ShadeEntity(plane, shadow)
ShadeEntity(anothercube, shadow)

tex = LoadTexture("media/test.jpg")

EntityTexture(anothercube, tex)
EntityTexture(plane,tex)

While Not KeyHit(key_escape) And Not AppTerminate()

	
	UpdateShadows()

	If KeyDown(KEY_W) Then MoveEntity(camera, 0, 0, 1)
	If KeyDown(KEY_S) Then MoveEntity(camera, 0, 0, -1)
	If KeyDown(KEY_A) Then MoveEntity(camera, 0, -1, 0)
	If KeyDown(KEY_D) Then MoveEntity(camera, 0, 1, 0)
	PointEntity(camera,cube)

	TurnEntity(cube, 1, 1, 1)

       
	UpdateWorld()
	RenderWorld()
	Flip()

Wend
```

**Comment by blitzcoder** [comment]
Ok got it. I haven't fully checked out the ob3dplus shadowmap version, but usually they are straightforward.

still not sure if this is also the case as shadow shaders are a bit complex to setup.

Yes, the shadow texture resolution is supposed to be a quick change on the shader parameters if this has been setup properly.

>OT: I can't actually test this as I lost my laptop and haven't had time to set up my new one, Windows 10 has been very slow on my new laptop and I'm not sure what to do, I've tried disabling Cortana and all the bloatware including Aero but it's still very slow, have you any suggestions there? Thanks.

try using a win10 debloater on a clean install, here's one: https://github.com/Sycnex/Windows10Debloater

**Comment by blitzcoder** [comment]
I'm also getting the same results directly using C, but both are shaded black.

![Untitled-1](https://user-images.githubusercontent.com/57479219/185728063-1f61072e-d2e0-4828-882f-d54db3302747.png)

If you checked out the latest OpenB3D Plus (0.12), the `CreateTexture flag (1+16+32)` is not set as per that resolved comment 

Here's a copy a shadow.cpp:
https://gist.github.com/blitzcoder/a3e04eee87f18d2d31c2f70f37de75fe

The Vortex2 shadowmap shader seems to work and it is in GLES2, it is minified though and combined with other shaders.

**Comment by markcwm** [comment]
Yes Openb3d doesn't get updated regularly, maybe it will be in the next version.

I haven't got a working install of Openb3d with GL 3 to test at the moment, but give me a while and I'll have a look.

**Comment by blitzcoder** [comment]
That's great. It would also be nice though if angros47 will include fully working examples in OB3D Plus so this can be tested and confirmed working in the first place.

**Comment by markcwm** [comment]
OK, trying to install on Mac as my Windows GCC is broken at the moment. 

This might be an issue with vanilla/OS Blitzmax, is that what you used to test? I only used NG to test and it worked OK and I didn't get the multiple shadows like Kronos did.

**Comment by markcwm** [comment]
Yes, it looks like it is an issue with vanilla/OS Blitzmax, as I get a MAV type error at LoadShader in shadowmap.bmx. I can't really see why it would do this though, I tried commenting out things but doing that could create another MAV. I have GL 3.3 which is more recent than 140 (GL 3.1) so it should work.

I haven't been able to build NG yet as it errors out, trying to build it with vanilla/OS Blitzmax didn't work either, so I guess I'll have to find an older version which is still compatible with vanilla/OS bmk.

**Comment by markcwm** [comment]
Well I've another version uploaded d322d8c6d36bb445667b0f7909701a106325eabf but it's not working, the main reason is it needs the inverse function and that wasn't available in my GL, it seems it's common for something to just not be there, especially in GLES. So the MAV was caused by several things, the f in 1.0f, the ; at the end of the main brackets, the #version 140, and the inverse function.

I copied an inverse function from somewhere but it just renders garbage, it must be the problem as that's the only code changed, so change the name inversemat4 to just inverse and as long as you have that it should work.

**Comment by markcwm** [comment]
Oh, in the example I forgot to reset the graphics resolution to 640x480 and the mapsize to 512. Also, quite confusing to me the plane just disappeared, it seemed to be inside the cube, so I changed the code a bit.

**Comment by markcwm** [comment]
I tried porting the getinverse function from Openb3d but it gives the same results so it looks like it may be a bug on Mac. 

So I'll set up in Windows and see, as I have just installed 8.1 and the malware I had on my secondhand laptop has gone.

**Comment by blitzcoder** [comment]
Good to hear. 

Btw, I got GL 4.3 on my Win10 desktop.

>This might be an issue with vanilla/OS Blitzmax, is that what you used to test?

Yes, also tested it on DX.

**Comment by markcwm** [comment]
OK, I've updated it again, it should work for you now. I tested in Bmax OS, Win 8.1, GL 4.3 but there is a rendering bug that I don't think I saw before, they're shadow strips moving away from the camera. 

Since this won't work on my fairly modern Mac with Xcode 7.2, GL 3.3 I won't be working on this any further.

**Comment by blitzcoder** [comment]
Yes, seems to be working now with that strip or bias which I think is normal in this setup.. but funny or weird how it only works and stuck at 640x480 at 512 shadow resolution.

Then again perhaps this is an older method as with newer techniques you can change texture size and/or resolution.

![sm](https://user-images.githubusercontent.com/57479219/191137886-8645f57b-fe65-44a2-8ef9-a067576478cc.png)

**Comment by blitzcoder** [comment]
just noticed and found another issue where the `CreatePlane` entity texture used here should normally scroll like in the example, for this it doesn't and is somehow stuck.

**Comment by blitzcoder** [comment]
I had a second look and did not first notice that you are using a camera viewport for the light and shadow camera.

So for using different shadow texture size, you just use the resolution height as the shadow texture size.

```
Graphics3D 1024,768,0,2
Local mapsize%=GraphicsHeight()
```
same as the old method of doing shadows, as long as there is a resolution for it, it works with the resolution height.

The plane issue I mentioned previously also might have something to do with the texturing. Perhaps try using `VertexTexCoords` instead of [PositionTexture](https://github.com/markcwm/openb3d.mod/blob/master/openb3dcore.mod/inc/TMesh.bmx#L382) as angros suggested with this plane solution. So it seems that you also have moved away from the default openb3d plane function which seems jittery and view angle issues.

Here's a test with desktop resolution of 1920x1080 
![1920](https://github.com/markcwm/openb3d.docs/assets/57479219/73428472-9e4c-490a-b995-60ad2b2d037e)
